### PR TITLE
Differentiate between pure and impure intrinsics and externs

### PR DESCRIPTION
**Pull request by abadams**
Various places in the compiler need to know if they can move around,
coalesce, or duplicate call nodes, and they each seem to use a different
heuristic. This PR adds call types that explicitly tag call nodes as
pure. Other calls are assumed to impure.

Also added an is_intrinsic helper that I've been meaning to do for ages,
to simplify checking if a call is a particular intrinsic.

A lot of the noise in the PR is just reindenting code.

### CODE REVIEW

**Comment by abadams** [comment]
The motivation for this is the store forwarding work, which is much more aggressive about lifting expressions out of loops, so it needs a clear idea of whether calls are pure.

**Comment by abadams** [comment]
@dsharlet @zvookin any object to me merging this? It's very prone to merge conflicts, so if it's going in I'd like to get it in sooner rather that later.

File: src/IR.h [hunk 1]
```diff
@@ -362,7 +362,7 @@ struct Evaluate : public StmtNode<Evaluate> {
 struct Call : public ExprNode<Call> {
     std::string name;
     std::vector<Expr> args;
-    typedef enum {Image, Extern, Halide, Intrinsic} CallType;
+    typedef enum {Image, Extern, PureExtern, Halide, Intrinsic, PureIntrinsic} CallType;
```
Comment (zvookin) [inline src/IR.h]: Might be worth adding a comment here. In particular I'd like to record that the Pure is a property of the thing being called, not the call site. I.e. there should never be two calls to the same function, one of which is Pure and other not. (Though enforcing this for PureExtern is difficult.) I'm not sure this invariant is all that critical to optimization, but I believe it is the intent of the IR design and thus should be documented.

**Comment by zvookin** [comment]
LGTM. A lot of this looks mechanical. I'd like to come up with a mechanism for creating the intrinsic calls that is a little less repetitive, but that can happen another time.

**Comment by dsharletg** [comment]
Sorry for the delay, I'm on a weird sleeping pattern with this cold. This looks good to me, I'll merge it.

**Pull request merged**